### PR TITLE
Correctly resolve Inherent Associated Types

### PR DESCRIPTION
**Pull request by fee1-dead**
I don't know if this is the best way to do this, but at least it is one way.

### CODE REVIEW

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -1396,7 +1396,28 @@ fn clean_qpath<'tcx>(hir_ty: &hir::Ty<'tcx>, cx: &mut DocContext<'tcx>) -> Type
                 ty::Projection(proj) => Res::Def(DefKind::Trait, proj.trait_ref(cx.tcx).def_id),
                 // Rustdoc handles `ty::Error`s by turning them into `Type::Infer`s.
                 ty::Error(_) => return Type::Infer,
-                _ => bug!("clean: expected associated type, found `{:?}`", ty),
+                // Otherwise, this is an inherent associated type.
+                _ => {
+                    // the compiler only resolves inherent associated types correctly in MIR,
```
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1401]:
```suggestion
                    // the compiler only resolves inherent associated types correctly after HIR,
```
astconv and type-checking happen much earlier than MIR.

File: src/librustdoc/clean/mod.rs [hunk 2]
```diff
@@ -1396,7 +1396,28 @@ fn clean_qpath<'tcx>(hir_ty: &hir::Ty<'tcx>, cx: &mut DocContext<'tcx>) -> Type
                 ty::Projection(proj) => Res::Def(DefKind::Trait, proj.trait_ref(cx.tcx).def_id),
                 // Rustdoc handles `ty::Error`s by turning them into `Type::Infer`s.
                 ty::Error(_) => return Type::Infer,
-                _ => bug!("clean: expected associated type, found `{:?}`", ty),
+                // Otherwise, this is an inherent associated type.
+                _ => {
```
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1400]: Why is this branch required? Why doesn't `hir_ty_to_ty` work?

**Comment by wesleywiser** [comment]
I don't think I have enough context to do give a good review here. 

r? @cjgillot 

Feel free to reroll if you didn't want to take over the review 🙂

File: src/librustdoc/clean/mod.rs (hunk 2 above)
Comment (fee1-dead) [reply src/librustdoc/clean/mod.rs:1400]: Because apparently there is no way to convert a MIR ty to a cleaned Type in rustdoc. But I might be wrong..

Comment (cjgillot) [reply src/librustdoc/clean/mod.rs:1400]: That's what `clean_middle_ty` does, see the previous arm `QPath::Resolved`.

**Comment by BoxyUwU** [comment]
I am not sure this is what we want, it seems like closer to what we want is for assoc types to stay unresolved into typeck so that for example: `<_>::Assoc` can work :thinking: Probably fine to land this so the feature is Useable Now but it would be good to record this somewhere so it doesnt get forgotten before stabilization.

**Comment by fmease** [comment]
> `<_>::Assoc` can work

Out of curiosity, can it ever work? Are there cases where it's possible to infer `_` just given a projection? Projections generally aren't injective. Since we fully support inherent associated *constants*, are there some programs where we can successfully infer the type & the trait just given `<_>::ASSOC` (where we are in the value namespace)? I could be missing something.

**Comment by BoxyUwU** [comment]
projections arent injective but we still relate the substs, so `<_>::Assoc eq <u32>::Assoc` could plausibly constrain `_ eq u32`

**Pull request merged**